grails-app/services/com/unifina/service/ProductService.groovy

```diff
@@ -17,27 +17,24 @@ class ProductService {
 	Random random = ThreadLocalRandom.current()
 
 	List<Product> relatedProducts(Product product, int maxResults, SecUser user) {
-		if (product == null) {
-			return new ArrayList<Product>()
-		}
 		// find Product.owner's other products
-		ListParams params = new ProductListParams(productOwner: product.owner, max: maxResults)
+		ListParams params = new ProductListParams(productOwner: product.owner, max: maxResults, publicAccess: true)
 		Set<Product> all = new HashSet<Product>(list(params, user))
 
 		// find other products from the same category
-		params = new ProductListParams(categories: [product.category].toSet(), max: maxResults)
-		Set<Product> cat = new HashSet<Product>(list(params, user))
-		all.addAll(cat)
+		params = new ProductListParams(categories: [product.category].toSet(), max: maxResults, publicAccess: true)
+		all.addAll(list(params, user))
 
-		all.removeIf { Product p -> p.id == product.id }
-		if (all.size() <= maxResults) {
-			return new ArrayList<Product>(all)
-		}
-		while (all.size() > maxResults) {
-			int i = random.nextInt(all.size() - 1)
-			all.remove(i)
```

Comment:
This isn't removing by index, but instead trying to remove `Integer i` from the Set. Thus the set never decreases in size.

Reply:
👍  I was thinking about a list. just missed the conversion from set to list